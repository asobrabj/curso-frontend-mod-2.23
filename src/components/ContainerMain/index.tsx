import Anotations from '../Anotations'
import Sidebar from '../Sidebar'
import * as S from './styles'

interface ITheme {
  isDark: boolean
  changeTheme: () => void
}

export default function ContainerMain({ changeTheme, isDark }: ITheme) {
  return (
    <S.Container className="container">
      <Sidebar changeTheme={changeTheme} isDark={isDark} />
      <Anotations />
    </S.Container>
  )
}
