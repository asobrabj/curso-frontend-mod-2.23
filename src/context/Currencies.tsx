import { createContext, useContext, useState } from 'react';
import instance from '../service/axiosaInstace';

export interface ICurrencyContext {
  to: string;
  setTo: React.Dispatch<React.SetStateAction<string>>;
  from: string;
  setFrom: React.Dispatch<React.SetStateAction<string>>;
  amount: string;
  setAmount: React.Dispatch<React.SetStateAction<string>>;
  fecthConversion: (
    to: string,
    from: string,
    amount: string,
  ) => Promise<string>;
  error: boolean;
}

interface ICurrencyProvider {
  children: React.ReactNode;
}

const CurrencyContext = createContext<ICurrencyContext>({} as ICurrencyContext);

export const CurrencyProvider: React.FC<ICurrencyProvider> = ({ children }) => {
  const [error, setError] = useState<boolean>(false);
  const [to, setTo] = useState<string>('');
  const [from, setFrom] = useState<string>('');
  const [amount, setAmount] = useState<string>('');

  const fecthConversion = async (
    to: string,
    from: string,
    amount: string,
  ): Promise<string> => {
    try {
      const response = await instance.get(`${to}/${from}/${amount}`);
      return response.data.conversion_result;
    } catch (err) {
      setError(true);
      return 'Erro na convers√£o';
    }
  };

  return (
    <CurrencyContext.Provider
      value={{
        fecthConversion,
        error,
        amount,
        from,
        to,
        setTo,
        setAmount,
        setFrom,
      }}
    >
      {children}
    </CurrencyContext.Provider>
  );
};

export const useCurrency = () => {
  return useContext(CurrencyContext);
};
