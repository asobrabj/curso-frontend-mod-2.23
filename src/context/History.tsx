/* eslint-disable react-refresh/only-export-components */
import { createContext, useContext, useEffect, useState } from 'react';

export interface IDataCurrency {
  to: string;
  from: string;
  amount: string;
  result: string;
  date: string;
}

export interface IHistoryContext {
  setHistory: (data: IDataCurrency) => void;
  history?: Array<IDataCurrency>;
}

interface IHistoryProvider {
  children: React.ReactNode;
}

const HistoryContext = createContext<IHistoryContext>({} as IHistoryContext);

export const HistoryProvider: React.FC<IHistoryProvider> = ({ children }) => {
  const [history, setHistoryState] = useState<Array<IDataCurrency>>();

  useEffect(() => {
    const currentData = window.localStorage.getItem('history');
    const newData = currentData ? JSON.parse(currentData) : [];
    setHistoryState(newData);
  }, []);

  const setHistory = (data: IDataCurrency) => {
    const currentData = window.localStorage.getItem('history');
    const newData = currentData ? JSON.parse(currentData) : [];
    newData.push(data);
    setHistoryState(newData);
    window.localStorage.setItem('history', JSON.stringify(newData));
    return;
  };

  return (
    <HistoryContext.Provider value={{ setHistory, history }}>
      {children}
    </HistoryContext.Provider>
  );
};

export const useHistory = () => {
  return useContext(HistoryContext);
};
