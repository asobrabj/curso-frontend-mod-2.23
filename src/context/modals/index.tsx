import { createContext, useContext, useState } from 'react'
import IAnotation from '../../types/anotation'

export interface IModalContext {
  isVisible: boolean
  setIsVisible: React.Dispatch<React.SetStateAction<boolean>>
  isForm: boolean
  setIsForm: React.Dispatch<React.SetStateAction<boolean>>
  dataCard: IAnotation
  setDataCard: React.Dispatch<React.SetStateAction<IAnotation>>
  idCard: number | null
  setIdCard: React.Dispatch<React.SetStateAction<number | null>>
}

interface IModalProvider {
  children: React.ReactNode
}

const ModalContext = createContext<IModalContext>({} as IModalContext)

export function ModalProvider({ children }: IModalProvider) {
  const [isVisible, setIsVisible] = useState<boolean>(false)
  const [isForm, setIsForm] = useState<boolean>(false)
  const [idCard, setIdCard] = useState<number | null>(null)
  const [dataCard, setDataCard] = useState<IAnotation>({
    id: null,
    concluded: false,
    date: '',
    description: '',
    name: '',
    category: '',
    priority: '',
    time: '',
  })

  return (
    <ModalContext.Provider
      value={{
        idCard,
        setIdCard,
        isVisible,
        setIsVisible,
        isForm,
        setIsForm,
        dataCard,
        setDataCard,
      }}
    >
      {children}
    </ModalContext.Provider>
  )
}

export function useModal() {
  return useContext(ModalContext)
}
