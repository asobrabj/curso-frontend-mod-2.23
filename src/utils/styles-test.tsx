import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
