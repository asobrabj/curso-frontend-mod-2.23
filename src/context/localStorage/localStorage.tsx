import { createContext, useContext, useEffect, useState } from 'react'
import { IFormData } from '../../components/FormAddAnnotation'
import IList from '../../types/lists'

export interface ILocalStorageContext {
  anotations: Array<IFormData> | null
  setAnotations: (data: IFormData | Array<IFormData>) => void
  getFilters: (value: string) => void
  getTags: (key: 'category' | 'priority', value: string) => void
  getStorage: (key: string) => Array<IFormData | IList> | string | null
  setStorage: (key: string, data: Array<IFormData | IList> | string) => void
}

interface ILocalStorageProvider {
  children: React.ReactNode
}

const LocalStorageContext = createContext<ILocalStorageContext>(
  {} as ILocalStorageContext
)

export function LocalStorageProvider({ children }: ILocalStorageProvider) {
  const [anotations, setAnotations] = useState<Array<IFormData> | null>([])

  useEffect(() => {
    const anotations = getStorage('anotations')
    setAnotations(anotations)
  }, [])

  const getTags = (key: 'category' | 'priority', value: string) => {
    const anotations = getStorage('anotations')
    const data = anotations?.filter(
      (item: IFormData) =>
        item[key]?.toLowerCase().includes(value.toLowerCase())
    )

    if (data) setAnotations(data)
    return
  }

  const getFilters = (value: string) => {
    const anotations = getStorage('anotations')
    const name = anotations?.filter((item: IFormData) =>
      item.name.toLowerCase().includes(value.toLowerCase())
    )

    if (name) setAnotations(name)
    return
  }

  const setAnotationsFn = (data: IFormData | Array<IFormData>) => {
    const storage = getStorage('anotations')
    let newStorage: Array<IFormData> = []
    if (storage) {
      newStorage = [...storage, data]
    }

    if (data instanceof Array) {
      newStorage = data
    }

    setAnotations(newStorage)
    localStorage.setItem('anotations', JSON.stringify(newStorage))
  }

  const getStorage = (key: string) => {
    const storedData = localStorage.getItem(key)
    if (storedData) {
      return JSON.parse(storedData)
    }
    return []
  }

  const saveToStorage = (
    key: string,
    data: Array<IFormData | IList> | string
  ) => {
    localStorage.setItem(key, JSON.stringify(data))
    return
  }
  return (
    <LocalStorageContext.Provider
      value={{
        getTags,
        getFilters,
        anotations,
        setAnotations: setAnotationsFn,
        getStorage,
        setStorage: saveToStorage,
      }}
    >
      {children}
    </LocalStorageContext.Provider>
  )
}

export function useStorage() {
  return useContext(LocalStorageContext)
}
