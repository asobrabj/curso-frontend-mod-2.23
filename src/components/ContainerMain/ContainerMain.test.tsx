import { fireEvent, screen } from '@testing-library/react'
import ContainerMain from '.'
import { LocalStorageProvider } from '../../context/localStorage/localStorage'
import { ModalProvider } from '../../context/modals'
import { renderTheme } from '../../utils/styles-test'

jest.mock('../../context/localStorage/localStorage', () => {
  const originalModule = jest.requireActual(
    '../../context/localStorage/localStorage'
  )

  const mockUseStorage = jest.fn(() => ({
    ...originalModule,
    getFilters: jest.fn(),
  }))

  return {
    ...originalModule,
    useStorage: mockUseStorage,
  }
})
describe('ContainerMain Component', () => {
  it('deve renderizar corretamente', () => {
    const { container } = renderTheme(
      <ModalProvider>
        <LocalStorageProvider>
          <ContainerMain changeTheme={() => {}} isDark={false} />
        </LocalStorageProvider>
      </ModalProvider>
    )

    expect(container.querySelector('.container')).toBeInTheDocument()
    expect(screen.getByTestId('sidebar')).toBeInTheDocument()
    expect(screen.getByTitle('container-cards')).toBeInTheDocument()
  })

  it('deve chamar a função changeTheme ao ser clicado', () => {
    const changeThemeMock = jest.fn()
    renderTheme(<ContainerMain changeTheme={changeThemeMock} isDark={false} />)

    fireEvent.click(screen.getByTitle('change theme'))

    expect(changeThemeMock).toHaveBeenCalled()
  })
  it('snapshot', () => {
    const { container } = renderTheme(
      <ModalProvider>
        <ContainerMain />
      </ModalProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
