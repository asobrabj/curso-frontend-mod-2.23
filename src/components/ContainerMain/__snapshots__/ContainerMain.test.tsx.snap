// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContainerMain Component snapshot 1`] = `
.c14 {
  padding: 32px;
  margin: auto;
  z-index: 0;
}

.c15 {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.c16 {
  font-size: 150px;
  color: #999;
}

.c17 {
  font-size: 32px;
}

.c17 >span {
  font-style: italic;
}

.c17 >span:hover {
  cursor: pointer;
  text-decoration: underline;
}

.c13 {
  cursor: pointer;
  margin-left: 450px;
  margin-top: 5px;
  margin-right: 80px;
  position: relative;
  border: 1px solid #999;
  width: 32px;
  height: 32px;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.c13 >i {
  position: absolute;
  top: 6px;
  left: 6px;
  font-size: 18px;
  border-radius: 50%;
  color: #bdbbbb;
}

.c3 {
  padding: 16px;
  font-size: 14px;
  font-weight: 400;
  border-bottom: 1px solid #999;
  display: block;
  transition: opacity 1s ease-in-out;
}

.c4 {
  padding: 16px;
  font-size: 14px;
  font-weight: 400;
  border-bottom: 1px solid #999;
  display: none;
  transition: opacity 1s ease-in-out;
}

.c2 {
  text-transform: capitalize;
  cursor: pointer;
}

.c2 >.paragraph {
  display: flex;
  justify-content: center;
  font-size: 18px;
  font-weight: bold;
  padding: 16px 0;
  border-bottom: 1px solid #999;
}

.c2 >.paragraph >i {
  margin-right: 5px;
}

.c9 {
  position: absolute;
  top: 10px;
  right: 70px;
  display: flex;
  align-items: center;
  cursor: pointer;
  padding: 8px;
}

.c10 {
  font-weight: bold;
  margin-right: 8px;
}

.c11 {
  font-size: 24px;
  margin-top: 5px;
}

.c7 {
  margin-left: 120px;
  border: 1px solid;
  border-radius: 8px;
  width: 300px;
  padding: 8px;
  font-size: 16px;
  font-weight: bold;
}

.c7::placeholder {
  color: #999;
}

.c1 {
  border-right: 1px solid #999;
  width: 240px;
  height: 100%;
  text-align: center;
}

.c5 {
  display: none;
  position: relative;
  height: 150px;
  border-bottom: 1px solid #999;
}

.c5 >.c6 {
  margin-top: 10px;
  margin-left: 0;
  width: 200px;
}

.c5 >.c8 {
  margin: 10px 0;
  top: 50px;
  left: 20px;
  padding: 2px;
}

.c5 >.c12 {
  margin-left: 20px;
  margin-top: 60px;
}

.c0 {
  display: flex;
  height: 89vh;
}

@media (max-width: 820px) {
  .c13 {
    margin-left: 226px;
  }
}

@media (max-width: 820px) {
  .c5 {
    display: block;
  }
}

<div>
   
  <div
    class="c0 container"
  >
    <nav
      class="c1"
      data-testid="sidebar"
    >
      <div
        class=""
      >
        <ul
          class="c2"
        >
          <li
            class="c3 paragraph"
          >
            <i>
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
                />
              </svg>
            </i>
            categoria
          </li>
          <li
            class="c4"
          >
            finanças
          </li>
          <li
            class="c4"
          >
            comida
          </li>
          <li
            class="c4"
          >
            estudo
          </li>
          <li
            class="c4"
          >
            trabalho
          </li>
        </ul>
      </div>
      <div
        class=""
      >
        <ul
          class="c2"
        >
          <li
            class="c3 paragraph"
          >
            <i>
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
                />
              </svg>
            </i>
            prioridade
          </li>
          <li
            class="c4"
          >
            urgente
          </li>
          <li
            class="c4"
          >
            alta
          </li>
          <li
            class="c4"
          >
            media
          </li>
          <li
            class="c4"
          >
            baixa
          </li>
        </ul>
      </div>
      <div
        class="c5"
      >
        <input
          class="c6 c7"
          name="search"
          placeholder="filtrar anotações"
          type="text"
          value=""
        />
        <div
          class="c8 c9"
          title="new-anotation"
        >
          <p
            class="c10"
          >
            Nova Anotação
          </p>
          <i
            class="c11"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                id="Circle_Plus"
              >
                <g>
                  <path
                    d="M15,12.5H12.5V15a.5.5,0,0,1-1,0V12.5H9a.5.5,0,0,1,0-1h2.5V9a.5.5,0,0,1,1,0v2.5H15A.5.5,0,0,1,15,12.5Z"
                  />
                  <path
                    d="M12,21.932A9.934,9.934,0,1,1,21.932,12,9.944,9.944,0,0,1,12,21.932ZM12,3.065A8.934,8.934,0,1,0,20.932,12,8.944,8.944,0,0,0,12,3.065Z"
                  />
                </g>
              </g>
            </svg>
          </i>
        </div>
        <div
          class="c12 c13"
          data-testid="change theme"
          title="change theme"
        >
          <i
            title="icon theme"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M264 480A232 232 0 0132 248c0-94 54-178.28 137.61-214.67a16 16 0 0121.06 21.06C181.07 76.43 176 104.66 176 136c0 110.28 89.72 200 200 200 31.34 0 59.57-5.07 81.61-14.67a16 16 0 0121.06 21.06C442.28 426 358 480 264 480z"
              />
            </svg>
          </i>
        </div>
      </div>
    </nav>
    <main
      class="c14"
      title="container-cards"
    >
      <div
        class="c15"
        title="not-found"
      >
        <i
          class="c16"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 5v14H5V5h14m1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9zM11 7h6v2h-6V7zm0 4h6v2h-6v-2zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
            />
          </svg>
        </i>
        <p
          class="c17"
        >
          Nenhuma anotação foi encontrada
           
          <span>
            Click aqui para adicionar
          </span>
        </p>
      </div>
    </main>
  </div>
   
</div>
`;
